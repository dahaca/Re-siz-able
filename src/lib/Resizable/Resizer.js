import React, { useState, useEffect, useRef } from 'react'
import styled from 'styled-components'
import {
  element,
  number,
  bool,
  string,
  arrayOf,
  oneOf,
  object,
} from 'prop-types'
import generateStyles from './utils/generateStyles'

const DefaultHandle = styled.div`
  position: absolute;
  opacity: 0;
  user-select: none;

  ${({ position }) => generateStyles(position)};

  :hover {
    opacity: ${({ hideHandles }) => (hideHandles ? 0 : 0.7)};
  }

  :active {
    opacity: ${({ hideHandles }) => (hideHandles ? 0 : 1)};
    cursor: grabbing;
  }
`

const Wrapper = styled.div`
  position: relative;

  > :first-child {
    width: ${({ width }) => width && '100%'};
    height: ${({ height }) => height && '100%'};
  }
`

const Resizer = ({
  children,
  handles = ['left', 'right'],
  minWidth = 100,
  maxWidth = 600,
  minHeight = 100,
  maxHeight = 600,
  hideHandles,
  preserveRatio,
  customHandle,
  className,
}) => {
  const [width, setWidth] = useState()
  const [height, setHeight] = useState()
  const [mouseDown, setMouseDown] = useState(false)
  const [initialPosition, setInitialPosition] = useState(null)
  const [initialWidth, setInitialWidth] = useState()
  const [initialHeight, setInitialHeight] = useState()
  const childElement = useRef()
  const Handle = customHandle ? customHandle : DefaultHandle

  useEffect(() => {
    window.addEventListener('mousemove', handleMouseMove)
    window.addEventListener('mouseup', handleMouseUp)

    return () => {
      window.removeEventListener('mousemove', handleMouseMove)
      window.removeEventListener('mouseup', handleMouseUp)
    }
  }, [mouseDown])

  useEffect(() => {
    setWidth(childElement.current.getBoundingClientRect().width)
    setHeight(childElement.current.getBoundingClientRect().height)
  }, [])

  const handleMouseDown = handle => e => {
    setMouseDown(handle)
    setInitialPosition({ x: e.pageX, y: e.pageY })
    setInitialWidth(childElement.current.getBoundingClientRect().width)
    setInitialHeight(childElement.current.getBoundingClientRect().height)
  }

  const handleMouseUp = () => {
    if (mouseDown) {
      setMouseDown(null)
    }
  }

  const handleMouseMove = e => {
    if (!mouseDown) {
      return
    }

    const checkNewWidth = width => {
      if (width < minWidth) return minWidth
      if (width > maxWidth) return maxWidth

      return width
    }

    const checkNewHeight = height => {
      if (height < minHeight) return minHeight
      if (height > maxHeight) return maxHeight

      return height
    }

    const setDimentions = (deltaHeight, deltaWidth, preserveRatio) => {
      if (preserveRatio) {
        const delta = Math.max(deltaWidth, deltaHeight)

        setHeight(checkNewHeight(initialHeight + delta))
        setWidth(checkNewWidth(initialWidth + delta))

        return
      }

      setHeight(checkNewHeight(initialHeight + deltaHeight))
      setWidth(checkNewWidth(initialWidth + deltaWidth))
    }

    switch (mouseDown) {
      case 'right':
        setWidth(checkNewWidth(initialWidth + (e.pageX - initialPosition.x)))
        break

      case 'left':
        setWidth(checkNewWidth(initialWidth + (initialPosition.x - e.pageX)))
        break

      case 'top':
        setHeight(checkNewHeight(initialHeight + (initialPosition.y - e.pageY)))
        break

      case 'bottom':
        setHeight(checkNewHeight(initialHeight + (e.pageY - initialPosition.y)))
        break

      case 'top-left':
        setDimentions(
          initialPosition.y - e.pageY,
          initialPosition.x - e.pageX,
          preserveRatio
        )
        break

      case 'top-right':
        setDimentions(
          initialPosition.y - e.pageY,
          e.pageX - initialPosition.x,
          preserveRatio
        )
        break

      case 'bottom-right':
        setDimentions(
          e.pageY - initialPosition.y,
          e.pageX - initialPosition.x,
          preserveRatio
        )
        break

      case 'bottom-left':
        setDimentions(
          e.pageY - initialPosition.y,
          initialPosition.x - e.pageX,
          preserveRatio
        )
        break

      default:
        console.error('Invalid handle type!')
        break
    }
  }

  const renderChildWithRef = () => {
    try {
      const child = React.Children.only(children)

      return React.cloneElement(child, {
        ref: childElement,
      })
    } catch (err) {
      console.error('The Resizer component can have only one child!')
    }
  }

  return (
    <Wrapper
      width={width}
      height={height}
      style={{ width: width, height: height }}
      className={className}
    >
      {renderChildWithRef()}

      {handles.map(handle => (
        <Handle
          key={handle}
          position={handle}
          onMouseDown={handleMouseDown(handle)}
          hideHandles={hideHandles}
        />
      ))}
    </Wrapper>
  )
}

Resizer.propTypes = {
  children: element.isRequired,
  handles: arrayOf(
    oneOf([
      'right',
      'left',
      'top',
      'bottom',
      'top-left',
      'top-right',
      'bottom-right',
      'bottom-left',
    ])
  ),
  minWidth: number,
  maxWidth: number,
  minHeight: number,
  maxHeigh: number,
  hideHandles: bool,
  preserveRatio: bool,
  customHandle: object,
  className: string,
}

export default Resizer
