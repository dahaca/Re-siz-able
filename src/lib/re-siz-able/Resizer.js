import React from 'react'
import styled from 'styled-components'
import {
  element,
  number,
  bool,
  string,
  arrayOf,
  oneOfType,
  oneOf,
  object,
  func,
} from 'prop-types'
import generateStyles from './utils/generateStyles'
import useResizer from './useResizer'

const DefaultHandle = styled.div`
  position: absolute;
  opacity: 0;
  touch-action: none;
  user-select: none;

  ${({ position }) => generateStyles(position)};

  :hover {
    opacity: ${({ hideHandles }) => (hideHandles ? 0 : 0.7)};
  }

  :active {
    opacity: ${({ hideHandles }) => (hideHandles ? 0 : 1)};
    cursor: grabbing;
  }
`

const Wrapper = styled.div`
  position: relative;
  touch-action: none;

  > :first-child {
    width: ${({ width }) => width && '100%'};
    height: ${({ height }) => height && '100%'};
  }
`

const Resizer = ({
  children,
  handles = ['left', 'right'],
  minWidth = 100,
  maxWidth = 1000,
  minHeight = 100,
  maxHeight = 1000,
  margin,
  hideHandles,
  preserveRatio,
  customHandle,
  className,
}) => {
  const { elementRef, width, height, handlePointerDown } = useResizer({
    minWidth,
    maxWidth,
    minHeight,
    maxHeight,
    preserveRatio,
    margin,
  })

  if (handles.length === 0) {
    console.error(
      'You must include at least one handle name in the handles array!'
    )
  }

  const Handle = customHandle ? customHandle : DefaultHandle

  const renderChildWithRef = () => {
    try {
      const child = React.Children.only(children)
      return React.cloneElement(child, {
        ref: elementRef,
      })
    } catch (err) {
      console.error('The Resizer component must have exactly one child!')
    }
  }

  return (
    <Wrapper
      width={width}
      height={height}
      style={{ width: width, height: height }}
      className={className}
    >
      {renderChildWithRef()}

      {handles.map(handle => (
        <Handle
          key={handle}
          position={handle}
          onMouseDown={handlePointerDown(handle)}
          onTouchStart={handlePointerDown(handle)}
          hideHandles={hideHandles}
        />
      ))}
    </Wrapper>
  )
}

Resizer.propTypes = {
  children: element.isRequired,
  handles: arrayOf(
    oneOf([
      'right',
      'left',
      'top',
      'bottom',
      'top-left',
      'top-right',
      'bottom-right',
      'bottom-left',
    ])
  ),
  minWidth: number,
  maxWidth: number,
  minHeight: number,
  maxHeight: number,
  margin: number,
  hideHandles: bool,
  preserveRatio: bool,
  customHandle: oneOfType([object, element, func]),
  className: string,
}

export default Resizer
