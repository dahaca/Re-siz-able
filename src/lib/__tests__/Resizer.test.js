import React from 'react'
import { render } from 'enzyme'
import Resizer from '../re-siz-able'

describe('Test Resizer', () => {
  it('should render correctly', () => {
    expect(
      render(
        <Resizer>
          <div>blah</div>
        </Resizer>
      )
    ).toMatchSnapshot()
  })
})
