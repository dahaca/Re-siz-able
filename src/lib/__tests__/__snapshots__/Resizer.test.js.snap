// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Resizer should render correctly 1`] = `
<div
  class="sc-bwzfXH fbmalI"
>
  <div>
    blah
  </div>
  <div
    class="sc-bdVaJa ifPxOL"
  />
  <div
    class="sc-bdVaJa iFZMkA"
  />
</div>
`;
